### PR TITLE
Add org.swi_prolog.swipl

### PR DESCRIPTION
**Pull request by LyzardKing**
### Please confirm your submission meets all the criteria

SWI-Prolog is an open source (BSD-2) implementation of the Prolog language with many extensions. 
(SWI-)Prolog is a versatile language. It is being used for business rule validation, natural language processing, software verification, software refactoring, network configuration, security, robotics, reasoning in legal and medical domains, graph processing, machine learning (ILP, PLP), linked data (RDF), mathematical proofs, and much more.

- [x] Please describe your application briefly.
- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I have [built][build] and tested the submission locally.
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

**Comment by LyzardKing** [comment]
bot build org.swi_prolog.swipl

File: org.swi_prolog.swipl.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+  <id>org.swi_prolog.swipl</id>
+  
+  <name>SWI-Prolog</name>
+  <summary>ISO/Edinburgh-style Prolog interpreter</summary>
+  <url type="homepage">https://www.jabref.org</url>
```
Comment (hfiguiere) [inline org.swi_prolog.swipl.appdata.xml:7]: why is this not linking to swi-prolog.org ?

File: org.swi_prolog.swipl.desktop
Comment (hfiguiere) [inline org.swi_prolog.swipl.desktop:1]: this file should be submitted upstream for inclusion

File: org.swi_prolog.swipl.appdata.xml
Comment (hfiguiere) [inline org.swi_prolog.swipl.appdata.xml:1]: this file should be submitted upstream for inclusion

File: org.swi_prolog.swipl.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+runtime: org.kde.Platform
+runtime-version: "6.7"
+sdk: org.kde.Sdk
+command: swipl-win
+
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --share=network
+  - --filesystem=host
```
Comment (hfiguiere) [inline org.swi_prolog.swipl.yml:11]: why is host permission necessary?

File: org.swi_prolog.swipl.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        tag: V9.3.7
+        commit: 93ec6b480bacec44134ae7e48e9759dc49a5ec4a
+        x-checker-data:
+          type: git
+          tag-pattern: "^V(\\d\\.[\\d.]+)$"
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=PGO
+      - -DSKIP_SSL_TESTS=ON
+  - name: launcher
```
Comment (hfiguiere) [inline org.swi_prolog.swipl.yml:28]: this should be part of the main module (ie the one that build the app.

File: org.swi_prolog.swipl.png
Comment (hfiguiere) [inline org.swi_prolog.swipl.png:1]: icon should be submitted upstream.

Also minimum size is 128x128. And 57x57 isn't even a standard size.

File: org.swi_prolog.swipl.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    buildsystem: simple
+    build-commands:
+      - install -D -m0644 org.swi_prolog.swipl.png /app/share/icons/hicolor/128x128/apps/org.swi_prolog.swipl.png
```
Comment (hfiguiere) [inline org.swi_prolog.swipl.yml:31]: this is not the actual size of the icon.

File: org.swi_prolog.swipl.appdata.xml (hunk 1 above)
Comment (LyzardKing) [reply org.swi_prolog.swipl.appdata.xml:7]: This is a typo from a previous file I copied.
I'll fix it now.

**Comment by hfiguiere** [comment]
This is missing

> Please describe your application briefly.

File: org.swi_prolog.swipl.appdata.xml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  <metadata_license>MIT</metadata_license>
+  <project_license>MIT</project_license>
+
+  <developer id="tld.vendor">
```
Comment (hfiguiere) [inline org.swi_prolog.swipl.appdata.xml:11]: this is an INVALID developer id.

File: org.swi_prolog.swipl.desktop
Comment (LyzardKing) [reply org.swi_prolog.swipl.desktop:1]: These files should be included upstream soon. I'm in contact with the upstream.

File: org.swi_prolog.swipl.appdata.xml
Comment (LyzardKing) [reply org.swi_prolog.swipl.appdata.xml:1]: as above

File: org.swi_prolog.swipl.png
Comment (LyzardKing) [reply org.swi_prolog.swipl.png:1]: As above. The icon upstream is currently 64x64

File: org.swi_prolog.swipl.yml (hunk 4 above)
Comment (LyzardKing) [reply org.swi_prolog.swipl.yml:31]: Will change

File: org.swi_prolog.swipl.appdata.xml (hunk 5 above)
Comment (LyzardKing) [reply org.swi_prolog.swipl.appdata.xml:11]: changed

File: org.swi_prolog.swipl.png
Comment (hfiguiere) [reply org.swi_prolog.swipl.png:1]: still too small. The linter will complain. (ie it will be blocked)

Comment (LyzardKing) [reply org.swi_prolog.swipl.png:1]: Updated the icon to 128x128

**Comment by LyzardKing** [comment]
The description is in the appdata file. Should it be elsewhere?

**Comment by LyzardKing** [comment]
If we want to add a screenshot to this repo, can we link to it in the appdata file?

File: org.swi_prolog.swipl.yml (hunk 2 above)
Comment (LyzardKing) [reply org.swi_prolog.swipl.yml:11]: being a programming language a more extended access to the system is typically needed. One example would be external drives.

This permission could be removed or replaced with something else once the flatpak works

**Comment by LyzardKing** [comment]
@hfiguiere I should have answered/fixed most of your points.
I noticed that there is an error with the screenshot.

**Comment by LyzardKing** [comment]
bot, build org.swi_prolog.swipl

File: org.swi_prolog.swipl.appdata.xml (hunk 5 above)
Comment (bbhtt) [reply org.swi_prolog.swipl.appdata.xml:11]: > changed

Doesn't seem to be fixed.

File: org.swi_prolog.swipl.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      (SWI-)Prolog is a versatile language. It is being used for business rule validation, natural language processing, software verification, software refactoring, network configuration, security, robotics, reasoning in legal and medical domains, graph processing, machine learning (ILP, PLP), linked data (RDF), mathematical proofs, and much more.
+    </p>
+  </description>
+  
+  <releases>
+  	<release version="9.3.6" date="2024-05-14"/>
+  </releases>
+  
+  <content_rating type="oars-1.0">
+  </content_rating>
```
Comment (bbhtt) [inline org.swi_prolog.swipl.appdata.xml:27]:
```suggestion
  <content_rating type="oars-1.1" />
```

**Comment by bbhtt** [comment]
Does this have any GUI? Why is this being submitted as an application and not a Sdk.Extension?

**Comment by LyzardKing** [comment]
@bbhtt Everything that is not strictly related to the flatpak will go upstream soon.
This has a qt terminal interface, and that is what is visible in the screenshot.
In the future it might be that this is split in two parts, an extension and the GUI.

**Comment by bbhtt** [comment]
> This has a qt terminal interface, and that is what is visible in the screenshot.

So this should be type=console-application then?

**Comment by LyzardKing** [comment]
This has its own ui, app menus, and it isn't running in a generic teminal.

What is the difference in the flatpak context?

**Comment by bbhtt** [comment]
bot, build org.swi_prolog.swipl

**Comment by bbhtt** [comment]
> This has its own ui, app menus, and it isn't running in a generic teminal.

Can you please add a provides tag with a binary child tag  https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#provides ?

I can check the build them. Properly tagging it as a console or desktop app is necessary to skip some automated checks.

File: org.swi_prolog.swipl.appdata.xml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  	<release version="9.3.6" date="2024-05-14"/>
+  </releases>
+  
+  <content_rating type="oars-1.1" />
+  
+  <launchable type="desktop-id">org.swi_prolog.swipl.desktop</launchable>
+
+  <screenshots>
+    <screenshot type="default">
+      <image>https://raw.githubusercontent.com/flathub/org.swi_prolog.swipl/swipl-win.png</image>
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply org.swi_prolog.swipl.appdata.xml]: the repository `flathub/org.swi_prolog.swipl` doesn't exist, and even then I'm pretty sure this references nothing.

[the remark this answers is not in the document]
Comment (LyzardKing) [reply org.swi_prolog.swipl.appdata.xml]: the screenshot is in the PR repo. I temporarily changed the url to the fork

**Comment by LyzardKing** [comment]
@bbhtt thanks for the help.
Is a console app meant to be run in a terminal? If that is the case then we should consider this a desktop app.

File: org.swi_prolog.swipl.yml (hunk 3 above)
Comment (hfiguiere) [reply org.swi_prolog.swipl.yml:28]: this wasn't addressed.

**Comment by LyzardKing** [comment]
bot, build org.swi_prolog.swipl

**Comment by LyzardKing** [comment]
bot, build org.swi_prolog.swipl

**Comment by bbhtt** [comment]
> Is a console app meant to be run in a terminal? If that is the case then we should consider this a desktop app.

Yea this should be a desktop-application. The screenshot made me think it was running inside a terminal.

**Comment by bbhtt** [comment]
https://github.com/flathub/flathub/pull/5332/files#r1641831956 isn't fixed.

**Comment by LyzardKing** [comment]
@bbhtt I'm not sure what I should change for that.
While the launcher parts (icon, appinfo, ...) are not upstream I kept them in a different part.
Other flatpak packages do the same.

Is there a different way?

**Comment by bbhtt** [comment]
Merge the sources array of the launcher module into the first one. Then add the `install` commands in `post-install:` in the first module.

**Comment by LyzardKing** [comment]
bot, build org.swi_prolog.swipl

**Comment by LyzardKing** [comment]
> Merge the sources array of the launcher module into the first one. Then add the `install` commands in `post-install:` in the first module.

Done.
If/when this is merged we need to change the screenshot url to the flathub repo instead of my fork

**Comment by LyzardKing** [comment]
@barthalion @bbhtt 
Thanks.
I'll update the screenshot url in the new repo.

**Pull request closed without merging**